src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1149,545 @@ struct Wallet
+    * param: address - main- or sub-address to get the index from
+    * return: [major_index, minor_index] if succeeded
+    * note: sets status error on fail
+    */
+    virtual std::pair<std::uint32_t, std::uint32_t> getSubaddressIndex(const std::string &address) const = 0;
+    /**
+    * brief: freeze - freeze enote "so they don't appear in balance, nor are considered when creating a transaction, etc." (https://github.com/monero-project/monero/pull/5333)
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * note: sets status error on fail
+    */
+    virtual void freeze(std::size_t idx) = 0;
+    virtual void freeze(const std::string &key_image) = 0;
+    // QUESTION : Should we add the option to freeze by tx_id? Meaning: freeze all enotes that we received by a single transaction.
+//    virtual void freeze(const std::string &tx_id) = 0;
+    /**
+    * brief: thaw - thaw enote that is frozen, so it appears in balance and can be spent in a transaction
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * note: sets status error on fail
+    */
+    virtual void thaw(std::size_t idx) = 0;
+    virtual void thaw(const std::string &key_image) = 0;
+    // QUESTION : Should we add the option to thaw by tx_id? Meaning: thaw all enotes that we received by a single transaction.
+//    virtual void thaw(const std::string &tx_id) = 0;
+    /**
+    * brief: isFrozen - check if enote is frozen
+    * param: idx - index of enote in `m_transfers`
+    * param: key_image - key image of enote
+    * QUESTION : Which approach should we use? We probably don't need both. Or are there other suggestions? (See WalletImpl::isFrozen for implementation)
+    *   Approach 1 : Use PendingTransaction, which just like tools::wallet2::multisig_tx_set has a vector of pending txs and an unordered set of signers pub keys.
+    *   Approach 2 : Use the encrypted hex string generated with tools::wallet2::save_multisig_tx(). This approach is already used in WalletImpl::restoreMultisigTransaction.
+    * __________ Approach 1 __________
+    * param: multisig_ptxs -
+    * __________ Approach 2 __________
+    * param: multisig_sign_data -
+    * ________________________________
+    * return : true if enote is frozen, else false
+    * note: sets status error on fail
+    */
+    virtual bool isFrozen(std::size_t idx) const = 0;
+    virtual bool isFrozen(const std::string &key_image) const = 0;
+    virtual bool isFrozen(const PendingTransaction &multisig_ptxs) const = 0;
+    virtual bool isFrozen(const std::string multisig_sign_data) const = 0;
+    // QUESTION : Should we add the following function, which returns strings containing the key image for every frozen enote? I think it can be useful
+//    virtual std::vector<std::string> getAllFrozenEnotes() const = 0;
+    // TODO : Would this better fit in one of `Subaddress`/`SubaddressAccount` classes?
```

Comment:
Can you give some possible arguments for pushing this down to the account level (if I understand you correctly)? I don't see any right now, but you have probably give this much more thought than I :)

Reply:
Quick update:
There is no need to move `createOneOffSubaddress` to the account level. Since this is an exotic function, maybe we should add a note/warning that this function should only be used in special cases?
(moneromoo: "It was intended to help people who generated a subaddress far into the random index values.")
e.g. something like:
```
* brief: createOneOffSubaddress - Create a new account or subaddress for given index, without adding it to known accounts and subaddresses.
*                                 Use `SubaddressAccount::addRow()` or `Subaddress::addRow()` instead, to make sure the wallet adds accounts/subaddresses consecutively and keeps track of accounts and subaddresses.
```

`WalletImpl::addSubaddress` & `WalletImpl::addSubaddressAccount` also do not refresh the known accounts/subaddresses, if that's not for a good reason (that I don't see) I would either add refresh at the end of those functions, or at least give the same warning as proposed above.
Seems the GUI only uses the `addRow` functions (see [here](https://github.com/search?q=repo%3Amonero-project/monero-gui%20addRow&type=code)) not the `addSubaddress*` ones (see [here](https://github.com/search?q=repo%3Amonero-project/monero-gui%20addSubaddress&type=code)).